src/reducers/versions.tsx

```diff
@@ -0,0 +1,218 @@
+import { Reducer } from 'redux';
+import { ActionType, createAction, getType } from 'typesafe-actions';
+
+type VersionId = number;
+
+type LocalizedString = {
+  [lang: string]: string;
+};
+
+type VersionCompatibility = {
+  [appName: string]: {
+    min: string;
+    max: string;
+  };
+};
+
+type VersionLicense = {
+  id: number;
+  isCustom: boolean;
+  name: LocalizedString;
+  text: LocalizedString;
+  url: string;
+};
+
+export type ExternalVersionEntry = {
+  binary: boolean | string;
+  depth: number;
+  directory: boolean;
+  filename: string;
+  mimetype: string;
+  modified: string;
+  path: string;
+  sha256: string;
+  size: number | null;
+};
+
+export type ExternalVersionFile = {
+  content: string;
+  created: string;
+  entries: {
+    [nodeName: string]: ExternalVersionEntry;
+  };
+  hash: string;
+  id: number;
+  is_mozilla_signed_extension: boolean;
+  is_restart_required: boolean;
+  is_webextension: boolean;
+  permissions: string[];
+  platform: string;
+  size: number;
+  status: string;
+  url: string;
+};
+
+export type ExternalVersion = {
+  channel: string;
+  compatibility: VersionCompatibility;
+  edit_url: string;
+  file: ExternalVersionFile;
+  has_been_validated: boolean;
+  id: VersionId;
+  is_strict_compatibility_enabled: boolean;
+  license: VersionLicense;
+  release_notes: LocalizedString | null;
+  reviewed: string;
+  url: string;
+  validation_url: string;
+  validation_url_json: string;
+  version: string;
+};
+
+type VersionFile = {
+  content: string;
+  created: string;
+  id: number;
+  size: number;
+};
+
+export enum VersionEntryType {
+  image,
+  binary,
+  text,
+  directory,
+}
+
+type VersionEntry = {
+  depth: number;
+  filename: string;
+  modified: string;
+  path: string;
+  type: VersionEntryType;
+};
+
+export type Version = {
+  entries: VersionEntry[];
+  id: VersionId;
+  reviewed: string;
+  version: string;
+};
+
+export const actions = {
+  loadVersionFile: createAction('LOAD_VERSION_FILE', (resolve) => {
+    return (payload: { filename: string; version: ExternalVersion }) =>
+      resolve(payload);
+  }),
+  loadVersionInfo: createAction('LOAD_VERSION_INFO', (resolve) => {
+    return (payload: { version: ExternalVersion }) => resolve(payload);
+  }),
+};
+
+export type VersionsState = {
+  versionInfo: {
+    [versionId: number]: Version;
+  };
+  versionFiles: {
+    [versionId: number]: {
+      [filename: string]: VersionFile;
```

Comment:
I don't think this will work because the filename is not unique. It should probably be an `id`, but I don't think it exists. Yet, `path` exists and it is unique, so we probably want to use that.

Reply:
You could use the sha256 hash for added uniqueness and easier serializability. But the path should work too.

Reply:
Thanks @EnTeQuAk. We want to use something that will be available in `VersionEntry`, and I don't think we have a need to store the `sha256` for our internal VersionEntries, so I think `path` makes the most sense.

Reply:
yeah, I like path too